### DISCUSSION TITLE
Nullable error when target type comes from external reference

### DISCUSSION THREAD

**Discussion by PaulVrugt** [riok/mapperly#679] [Q&A]
I created a test project (attached it), in which we are getting nullable errors, but only when the target of the mapping is from a different library. But I can't figure out why this happens. 

[local projects.zip](https://github.com/riok/mapperly/files/12430215/local.projects.zip)

We have a source:

```c#
public class Source
{
    public SubSource[]? Property1 { get; set; }
}

public class SubSource
{
    public Guid Id { get; set; }
}
```

and a mapper:

```c#
[Mapper]
internal static partial class EntityMapper
{
    internal static partial Target ToMutationIssue(Source issueDto);
}
```

and the target, but this target is included via a references project:
```c#
public class Target
{
    public SubTarget[]? Property1 { get; set; }
}

public class SubTarget
{
    public Guid Id { get; set; }
}
```

now this generates:

```c#
internal static partial class EntityMapper
    {
        internal static partial global::dtoTest.Target ToMutationIssue(global::NullableMapperlyTest.Source issueDto)
        {
            var target = new global::dtoTest.Target();
            if (issueDto.Property1 != null)
            {
                target.Property1 = MapToSubTargetArray(issueDto.Property1);
            }

            return target;
        }

        private static global::dtoTest.SubTarget MapToSubTarget(global::NullableMapperlyTest.SubSource source)
        {
            var target = new global::dtoTest.SubTarget();
            target.Id = source.Id;
            return target;
        }

        private static global::dtoTest.SubTarget[] MapToSubTargetArray(global::NullableMapperlyTest.SubSource[] source)
        {
            var target = new global::dtoTest.SubTarget[source.Length];
            for (var i = 0; i < source.Length; i++)
            {
                target[i] = MapToSubTarget(source[i]);
            }

            return target;

        }
    }
```

but it also results in the following nullable warning:
```
EntityMapper.g.cs(12,36,12,75): error CS8619: Nullability of reference types in value of type 'SubTarget?[]' doesn't match target type 'SubTarget[]'.
```

the weird thing is, if I move the `Target` class to the same project, the issue goes away.

**Comment by TimothyMakkison** [comment]
Thanks for the bug report  😊 

Do you know which version of mapperly you are using? I recently changed the way arrays work in ~3.1.0~ 3.0.0

**Reply by PaulVrugt** [reply]
We are using 3.1, but I've also tried version 3.0, since I found the change you are referring to. But 3.0 also has this issue.

I thought I could easily work around this issue, but it turns out this issue is blocking me from further implementing mapperly, since we treat all warnings as error, and I can't find a way to suppress this particular one only in the mapperly generated code. So this is a quite pressing issue for us

**Reply by TimothyMakkison** [reply]
I've been able to recreate the issue, both in your file and in a separate project. It also occurs in rider and in Mapperly 2.8.0.

There does seem to a difference in the generated code between VS and rider. VS doesn't show any nullable annotations whereas rider does. I have no idea why this happens.

```C#
private static global::SubTarget? MapToSubTarget(global::SubSource source)
{
    var target = new global::SubTarget();
    target.Id = source.Id;
    return target;
}

private static global::SubTarget?[] MapToSubTargetArray(global::SubSource[] source)
{
    var target = new global::SubTarget?[source.Length];
    for (var i = 0; i < source.Length; i++)
    {
        target[i] = MapToSubTarget(source[i]);
    }

    return target;
}
``` 

Thanks for your detailed bug report 🙏

**Comment by latonz** [comment]
Created https://github.com/riok/mapperly/issues/690
